[CL-909] Add banner image to home page model

nTraum opened the pull request:
When testing how the existing `AppHeaderBgUploader` works with the `HomePage`, I've noticed that the uploader is monkey-patched for EE to take into account the tenant ID for the storage directory. See `MultiTenancy::Patches::AppHeaderBgUploader` in EE for details.
The tricky part is that `HomePage` does not have a relation to tenant as `AppConfiguration` does.

As the `Tenant` model is deprecated anyway and we want to move away from it (See TEC-15 for details), I opted to introduce a new uploader `HeaderBgUploader` here. It functions exactly the same as the old `AppHeaderBgUploader`, but works with `AppConfiguration` instead of `Tenant` and therefore doesn't need the monkey-patching anymore, which makes me very happy!
I've added deprecation notes to the relevant classes, we can remove them once the data migration is complete. The uploader also has tests now.

EE PR: https://github.com/CitizenLabDotCo/citizenlab-ee/pull/258

nTraum committed:
[CL-909] Add banner image to home page

nTraum committed:
[CL-909] Introduce a new HeaderBgUploader that does not use deprecated Tenant

When testing how the existing `AppHeaderBgUploader` works with the `HomePage`, I've noticed that the uploader is monkey-patched for EE to take into account the tenant ID for the storage directory. See `MultiTenancy::Patches::AppHeaderBgUploader` in EE for details.
The tricky part is that `HomePage` does not have a relation to tenant as `AppConfiguration` does.

As the `Tenant` model is deprecated anyway and we want to move away from it (See TEC-15 for details), I opted to introduce a new uploader `HeaderBgUploader` here. It functions exactly the same as the old `AppHeaderBgUploader`, but works with `AppConfiguration` instead of `Tenant` and therefore doesn't need the monkey-patching anymore, which makes me very happy!
I've added deprecation notes to the relevant classes, we can remove them once the data migration is complete. The uploader also has tests now.

nTraum committed:
[CL-909] Fix header_bg_uploader file name

nTraum committed:
Fix Rubocop offenses

nTraum commented on back/app/uploaders/base_uploader.rb:
```diff
@@ -2,7 +2,7 @@
 class BaseUploader < CarrierWave::Uploader::Base
   def self.use_fog_engine?
-    return if Rails.env.test? || Rails.env.development?
+    return false if Rails.env.test? || Rails.env.development?
```
Question mark methods should return `false` or `true`, not `nil`.

nTraum commented:
Hey @adessy: Does my proposal of introducing a new `Uploader` that references `AppConfiguration` instead of `Tenant` make sense to you? We are saving ourselves the monkeypatching, which felt :sparkles:. Just want to make sure I haven't overseen anything.

nTraum commented on back/app/uploaders/header_bg_uploader.rb:
```diff
@@ -0,0 +1,21 @@
+# AppConfiguration id and model id.
+class HeaderBgUploader < BaseImageUploader
+  def store_dir
+    File.join('uploads', AppConfiguration.instance.id, 'header-background', model.id)
```
See https://github.com/citizenlabdotco/citizenlab-ee/blob/CL-909_add_banner_image_to_home_page/back/engines/ee/multi_tenancy/app/uploaders/multi_tenancy/patches/app_header_bg_uploader.rb#L8-L10 for comparison. Basically the same implementation, but based on the ID of `AppConfiguration` instead of `Tenant`.

adessy commented:
> Hey @adessy: Does my proposal of introducing a new `Uploader` that references `AppConfiguration` instead of `Tenant` make sense to you? We are saving ourselves the monkeypatching, which felt ✨. Just want to make sure I haven't overseen anything.

Yes, I think so. But I'm wondering if it could an occasion to make things simpler and if we could just use the `store_dir` implementation of the parent class. There must have been a (historical?) reason why those uploaders were treated differently, but I don't quite remember it. It was probably linked somehow to the particular role `Tenant` and `AppConfiguration` play. But since we're moving this to a "regular" model, maybe we can reuse the same logic as for other uploaders.

Little "did you know?" fact about `AppConfiguration`:  The `AppConfiguration` singleton has the same `id` as its `Tenant`. I'm not sure if everyone is aware of that 🙂

nTraum commented:
I'd be happy to just use the store dir of the parent class, but I'm not sure about the implications. The store dir was actually changed by you the last time(s), but I can't find a reason in the commits / PR, do you remember? :)

https://github.com/citizenlabdotco/citizenlab/commit/0dca0b6056c2e17d780dd4aa8cb0d2b42247ee43
https://github.com/citizenlabdotco/citizenlab/commit/94e42350eacf14b8b84a1e86f6bbd11574ca987d

94e42350eacf14b8b84a1e86f6bbd11574ca987d introduced the store dir on that class in the first place, but there is no ticket / PR attached.

jinjagit reviewed (approved):
I'm not very knowledgable on image encoding, etc. So best I can say is nothing looks obviously wrong to me.

adessy commented:
> I'd be happy to just use the store dir of the parent class, but I'm not sure about the implications. The store dir was actually changed by you the last time(s), but I can't find a reason in the commits / PR, do you remember? :)

😬🙈  I'm not sure if we can trust old me on this one. But I could find traces of a differentiated treatment between `store_dir` that predates those commits (in the old cl2-back repository). I'd say the main reason behind this difference was that in some cases (maybe via the admin API) the code of the uploader was run outside the context of a tenant. But I guess it should not be the case anymore if it's moved to this new `HomePage` model.

nTraum commented:
I see, I'll give it a try and remove the custom `store_dir` then. In this particular instance it won't have any effect anyway, as we save the file base64-encoded in the column anyway as far as I can see.

nTraum committed:
[CL-909] Remove custom `store_dir` for HeaderBgUploader

We don't need it anymore, see https://github.com/CitizenLabDotCo/citizenlab/pull/2151#issuecomment-1153653877 for details.

Also https://github.com/CitizenLabDotCo/citizenlab/pull/2151#issuecomment-1154011221:

> I'm not sure if we can trust old me on this one. But I could find traces of a differentiated treatment between store_dir that predates those commits (in the old cl2-back repository). I'd say the main reason behind this difference was that in some cases (maybe via the admin API) the code of the uploader was run outside the context of a tenant. But I guess it should not be the case anymore if it's moved to this new HomePage model.
